File: java/src/main/java/br/com/emendes/problems/SearchInRotatedSortedArrayII.java
package br.com.emendes.problems;

/**
 * Existe um array de inteiros nums ordenados em ordem não decrescente (não necessáriamente com valores distintos).<br><br>
 * <p>
 * Antes de passar pela sua função, nums é rotacionado em um índice pivô desconhecido.
 * Por exemplo, [0,1,2,4,4,4,5,6,6,7] pode ser rotacionado no índice pivô 5 e se tornar [4,5,6,6,7,0,1,2,4,4].<br><br>
 * <p>
 * Dado o array nums depois de rotacionado e um inteiro target, retorne true se target está em nums,
 * ou false caso contrário.<br><br>
 * <p>
 * Você deve diminuir ao máximo as etapas gerais da operação.<br><br>
 * <p>
 * Restrições:<br>
 * 1 <= nums.length <= 5000<br>
 * -10⁴ <= nums[i] <= 10⁴<br>
 * É garantido que nums seja rotacionado em algum pivô.<br>
 * -10⁴ <= target <= 10⁴
 */
public class SearchInRotatedSortedArrayII {

  public boolean search(int[] nums, int target) {
    for (int num : nums) {
      if (num == target) return true;
    }

    return false;
  }

}

File: javascript/problems/SearchInRotatedSortedArrayII.js
/**
 * Tem um array de inteiros nums ordenado em ordem crescente (não necessáriamente de valores únicos).
 *
 * Antes de ser passado para sua função, nums é rotacionado em um indice pivo desconhecido
 * k (0 <= k < nums.length) de modo que o array fica
 * [nums[k], nums[k+1],..., nums[n-1], nums[0], nums[1],...,nums[k-1]].
 *
 * Dado um array nums depois da rotação e um inteiro target, retorne true se target esta em nums,
 * ou false caso contrário.
 *
 * Você deve reduzir as etapas gerais da operação o máximo possível.
 *
 * Restrições:
 *
 * 1 <= nums.length <= 5000
 *
 * -10⁴ <= nums[i] <= 10⁴
 *
 * É garantido que nums esta rotacionado em algum pivo.
 *
 * -10⁴ <= target <= 10⁴
 *
 * Follow-up: Esse problema é similar a Search in Rotated Sorted Array, mas nums pode conter duplicatas.
 * Poderia isso afetar a complexidade de tempo? Como e Por quê?
 */
const search = function (nums, target) {
  let left = 0;
  let right = nums.length - 1;
  while (left <= right) {
    if (nums[left] === nums[right]) {
      if (target === nums[left]) return true;
      left++;
      right--;
    } else {
      let mid = Math.floor((right + left) / 2);
      if (target === nums[mid]) return true;
      if (nums[left] <= nums[mid]) {
        if (target >= nums[left] && target < nums[mid]) right = mid - 1;
        else left = mid + 1;
      } else if (target > nums[mid] && target <= nums[right]) left = mid + 1;
      else right = mid - 1;
    }
  }

  return false;
};
